packages/opencensus-instrumentation-grpc/src/grpc.ts

```diff
@@ -0,0 +1,478 @@
+          GrpcPlugin.ATTRIBUTE_GRPC_STATUS_CODE, call.status.code.toString());
+      // if there is an error, span is ended on error event, otherwise here
+      if (call.status.code === 0) {
+        endSpan();
+      }
+    });
+
+    call.on('error', (err: grpcModule.ServiceError) => {
+      rootSpan.addAttribute(GrpcPlugin.ATTRIBUTE_GRPC_ERROR_NAME, err.name);
+      rootSpan.addAttribute(
+          GrpcPlugin.ATTRIBUTE_GRPC_ERROR_MESSAGE, err.message);
+      endSpan();
+    });
+
+    return original.call(self, call);
+  }
+
+  /**
+   * Returns a function that patches the gRPC makeClientConstructor in order
+   * to patch all client methods.
+   * @returns {Function}  The register patched function.
+   */
+  private patchClient() {
+    const plugin = this;
+    return (original: MakeClientConstructor) => {
+      plugin.logger.debug('patchClient');
+      return function makeClientConstructor(
+          // tslint:disable-next-line:no-any
+          methods: grpcModule.ServiceDefinition<any>, serviceName: string,
+          options: grpcModule.GenericClientOptions) {
+        const client = original.apply(this, arguments);
+        shimmer.massWrap(
+            client.prototype, Object.keys(methods) as never[],
+            plugin.patchAllClientsMethods());
+        return client;
+      };
+    };
+  }
+
+
+  /**
+   * This function starts a span (child or root) immediately before the
+   * client method is invoked, and ends it either in a callback or stream
+   * event handler, depending on the method type.
+   */
+  private patchAllClientsMethods() {
+    const plugin = this;
+    // tslint:disable-next-line:no-any
+    return (original: any) => {
+      plugin.logger.debug('patchAllClientsMethods');
+      return function clientMethodTrace() {
+        const traceOptions = {
+          name: `grpc.${original.path.replace('/', '')}`,
+          type: 'CLIENT',
+        };
+        const args = Array.prototype.slice.call(arguments);
+        // Checks if this remote function call is part of an operation by
+        // checking if there is a current root span, if so, we create a child
+        // span. In case there is no root span, this means that the remote
+        // function call is the first operation, therefore we create a root
+        // span.
+        if (!plugin.tracer.currentRootSpan) {
+          return plugin.tracer.startRootSpan(
+              traceOptions,
+              plugin.makeGrpcClientRemoteCall(original, args, this, plugin));
+        } else {
+          const span = plugin.tracer.startChildSpan(
+              traceOptions.name, traceOptions.type);
+          return (plugin.makeGrpcClientRemoteCall(
+              original, args, this, plugin))(span);
+        }
+      };
+    };
+  }
+
+  /**
+   * This method handels the client remote call
+   */
+  private makeGrpcClientRemoteCall(
+      // tslint:disable-next-line:no-any
+      original: any, args: any[], self: any, plugin: GrpcPlugin) {
```

Comment:
Can you specify the actual types here?

Reply:
`self: Client`
I think `original` can be Function `args` can be any for now, diving into types here doesn't help us.

Reply:
Thanks. Done!

Reply:
Done!